src/net.rs

```diff
@@ -236,6 +226,64 @@ impl Gossip {
     }
 }
 
+/// Stream of events for a topic.
+#[derive(derive_more::Debug)]
+pub struct EventStream {
+    /// The actual stream polled to return [`Event`]s to the application.
+    #[debug("Stream")]
+    inner: Pin<Box<dyn Stream<Item = Result<Event>> + Send + 'static>>,
+
+    /// Channel to the actor task.
+    ///
+    /// This is used to handle the receiver being dropped. When all receiver and publishers are
+    /// gone the topic will be desubscribed.
+    to_actor_tx: mpsc::Sender<ToActor>,
+    /// The topic for which this stream is reporting events.
+    ///
+    /// This is sent on drop to the actor to handle the receiver going away.
+    topic: TopicId,
+    /// An Id identifying this specific receiver.
+    ///
+    /// This is sent on drop to the actor to handle the receiver going away.
+    receiver_id: ReceiverId,
+}
+
+impl Stream for EventStream {
+    type Item = Result<Event>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        self.inner.poll_next(cx)
+    }
+}
+
+impl Drop for EventStream {
+    fn drop(&mut self) {
+        // NOTE: unexpectedly, this works without a tokio runtime, so we leverage that to avoid yet
+        // another spawned task
+        if let Err(e) = self.to_actor_tx.try_send(ToActor::ReceiverGone {
+            topic: self.topic,
+            receiver_id: self.receiver_id,
+        }) {
```

Comment:
You could actually use [`to_actor_tx.reserve_owned`](https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Sender.html#method.reserve_owned) and store the `OwnedPermit` in `EventStream`.

This way you'll guarantee that you can always have space for this message and sending it will always be a synchronous call.

Reply:
I'm not sure this is a good idea. The actor channel is (correctly) bounded, with just a capacity of 64. So this would top our maximum subscriptions to that amount and block any other message to the actor since the permit is long lived (from subscription start to end)

Reply:
Good point.

Reply:
Something like a channel that is bounded for normal messages but unbounded for special shutdown messages would be nice sometimes...